core/domain/acl_decorators.py

```diff
@@ -72,40 +71,286 @@ def check_activity_accessible(
             activity_rights.viewable_if_private)
 
 
-def check_exploration_editable(user_id, user_actions, exploration_id):
-    """Returns a boolean to signify whether given exploration is editable
-    by the user or not.
+def check_can_edit_activity(
+        user_id, user_actions, activity_type, activity_id):
+    """Checks whether the user can edit given activity.
 
     Args:
         user_id: str. Id of the given user.
-        user_actions: list(str). List of actions given user can perform.
-        exploration_id: str. Exploration id.
+        user_actions: list(str). List of actions the user can perform.
+        activity_id: str. Id of the given activity.
 
     Returns:
-        bool. Whether the given exploration can be accessed.
+        bool. Whether the given user can edit this activity.
     """
+    if user_id is None:
+        return False
+
+    activity_rights = (
+        rights_manager.get_exploration_rights(
+            activity_id, strict=False)
+        if activity_type == feconf.ACTIVITY_TYPE_EXPLORATION
+        else rights_manager.get_collection_rights(activity_id, strict=False))
+
+    if activity_rights is None:
+        raise base.UserFacingExceptions.PageNotFoundException
+
+    action_edit_any_activity = (
+        role_services.ACTION_EDIT_ANY_EXPLORATION
+        if activity_type == feconf.ACTIVITY_TYPE_EXPLORATION
+        else role_services.ACTION_EDIT_ANY_COLLECTION)
+    action_edit_any_public_activity = (
+        role_services.ACTION_EDIT_ANY_PUBLIC_EXPLORATION
+        if activity_type == feconf.ACTIVITY_TYPE_EXPLORATION
+        else role_services.ACTION_EDIT_ANY_PUBLIC_COLLECTION)
+    action_edit_owned_activity = (
+        role_services.ACTION_EDIT_OWNED_EXPLORATION
+        if activity_type == feconf.ACTIVITY_TYPE_EXPLORATION
+        else role_services.ACTION_EDIT_OWNED_COLLECTION)
+
+    if action_edit_owned_activity in user_actions:
+        if (activity_rights.is_owner(user_id) or
+                activity_rights.is_editor(user_id)):
+            return True
+    else:
+        return False
+
+    if (activity_rights.community_owned or
+            (action_edit_any_activity in user_actions)):
+        return True
+
+    if (activity_rights.is_published() and
+            (action_edit_any_public_activity in user_actions)):
+        return True
+
+    return False
+
+
+def check_can_unpublish_collection(user_id, user_actions, collection_id):
+    """Checks whether the user can unpublish given collection.
+
+    Args:
+        user_id: str. Id of the user.
```

Comment:
Again I'm concerned about this since the user_id can be None. This really needs to be checked carefully for the other functions.